src/analysis/processing/qgsalgorithmreclassifybylayer.cpp

```diff
@@ -0,0 +1,298 @@
+
+  std::unique_ptr< QgsProcessingParameterBoolean > missingValuesParam = qgis::make_unique< QgsProcessingParameterBoolean >( QStringLiteral( "NODATA_FOR_MISSING" ),
+      QObject::tr( "Use no data when no range matches value" ), false, false );
+  missingValuesParam->setFlags( QgsProcessingParameterDefinition::FlagAdvanced );
+  addParameter( missingValuesParam.release() );
+
+  addParameter( new QgsProcessingParameterRasterDestination( QStringLiteral( "OUTPUT" ), QObject::tr( "Reclassified raster" ) ) );
+}
+
+bool QgsReclassifyAlgorithmBase::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  QgsRasterLayer *layer = parameterAsRasterLayer( parameters, QStringLiteral( "INPUT_RASTER" ), context );
+  mBand = parameterAsInt( parameters, QStringLiteral( "RASTER_BAND" ), context );
+
+  if ( !layer )
+    throw QgsProcessingException( invalidRasterError( parameters, QStringLiteral( "INPUT_RASTER" ) ) );
+
+  mInterface.reset( layer->dataProvider()->clone() );
+  mExtent = layer->extent();
+  mCrs = layer->crs();
+  mRasterUnitsPerPixelX = std::abs( layer->rasterUnitsPerPixelX() );
+  mRasterUnitsPerPixelY = std::abs( layer->rasterUnitsPerPixelX() );
+  mNbCellsXProvider = mInterface->xSize();
+  mNbCellsYProvider = mInterface->ySize();
+
+  mNoDataValue = parameterAsDouble( parameters, QStringLiteral( "NO_DATA" ), context );
+  mUseNoDataForMissingValues = parameterAsBool( parameters, QStringLiteral( "NODATA_FOR_MISSING" ), context );
+
+  int boundsType = parameterAsEnum( parameters, QStringLiteral( "RANGE_BOUNDARIES" ), context );
+  switch ( boundsType )
+  {
+    case 0:
+      mBoundsType = QgsReclassifyUtils::RasterClass::IncludeMax;
+      break;
+
+    case 1:
+      mBoundsType = QgsReclassifyUtils::RasterClass::IncludeMin;
+      break;
+
+    case 2:
+      mBoundsType = QgsReclassifyUtils::RasterClass::IncludeMinAndMax;
+      break;
+
+    case 3:
+      mBoundsType = QgsReclassifyUtils::RasterClass::Exclusive;
+      break;
+  }
+
+  return _prepareAlgorithm( parameters, context, feedback );
+}
+
+QVariantMap QgsReclassifyAlgorithmBase::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  QVector< QgsReclassifyUtils::RasterClass > classes = createClasses( mBoundsType, parameters, context, feedback );
+
+  const QString outputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT" ), context );
+  QFileInfo fi( outputFile );
+  const QString outputFormat = QgsRasterFileWriter::driverForExtension( fi.suffix() );
+
+  std::unique_ptr< QgsRasterFileWriter > writer = qgis::make_unique< QgsRasterFileWriter >( outputFile );
+  writer->setOutputProviderKey( QStringLiteral( "gdal" ) );
+  writer->setOutputFormat( outputFormat );
+  std::unique_ptr<QgsRasterDataProvider > provider( writer->createOneBandRaster( Qgis::Float32, mNbCellsXProvider, mNbCellsYProvider, mExtent, mCrs ) );
```

Comment:
Alternatively, the algorithm could cleverly choose the data type on the basis of the re-classified values.

if ( reclassified min >= 0 and reclassified max <= 255 ) byte
else if ( reclassified min  =>0 and reclassified max <= 65535) int16
etc. etc.

Reply:
But that assumes that the classes cover the whole range of input values - because if not then you'd also have to consider the range of input raster values